validator/db/kv/attester_protection.go

```diff
@@ -101,26 +110,102 @@ func (store *Store) CheckSlashableAttestation(
 func (store *Store) ApplyAttestationForPubKey(
 	ctx context.Context, pubKey [48]byte, signingRoot [32]byte, att *ethpb.IndexedAttestation,
 ) error {
-	return store.update(func(tx *bolt.Tx) error {
-		bucket := tx.Bucket(pubKeysBucket)
-		pkBucket, err := bucket.CreateBucketIfNotExists(pubKey[:])
+	store.batchedAttestationsChan <- &attestationRecord{
+		pubKey:      pubKey,
+		source:      att.Data.Source.Epoch,
+		target:      att.Data.Target.Epoch,
+		signingRoot: signingRoot,
+	}
+	// Subscribe to be notified when the attestation record queued
+	// for saving to the DB is indeed saved. If an error occurred
+	// during the process of saving the attestation record, the sender
+	// will give us that error. We use a buffered channel
+	// to prevent blocking the sender from notifying us of the result.
+	errChan := make(chan error, 1)
+	defer close(errChan)
+	sub := store.batchAttestationsFlushedFeed.Subscribe(errChan)
+	defer sub.Unsubscribe()
+	return <-errChan
+}
+
+// Meant to run as a background routine, this function checks whether:
+// (a) we have reached a max capacity of batched attestations in the Store or
+// (b) ATTESTATION_BATCH_WRITE_INTERVAL has passed
+// Based on whichever comes first, this function then proceeds
+// to flush the attestations to the DB all at once in a single boltDB
+// transaction for efficiency. Then, batched attestations slice is emptied out.
+func (store *Store) batchAttestationWrites(ctx context.Context) {
+	timer := time.NewTimer(ATTESTATION_BATCH_WRITE_INTERVAL)
+	defer timer.Stop()
+	for {
+		select {
+		case v := <-store.batchedAttestationsChan:
+			store.batchedAttestations = append(store.batchedAttestations, v)
+			if len(store.batchedAttestations) == ATTESTATION_BATCH_CAPACITY {
+				log.WithField("numRecords", ATTESTATION_BATCH_CAPACITY).Debug(
+					"Reached max capacity of batched attestation records, flushing to DB",
+				)
+				store.flushAttestationRecords()
+				timer.Reset(ATTESTATION_BATCH_WRITE_INTERVAL)
```

Comment:
> Reset should be invoked only on stopped or expired timers with drained channels.
https://pkg.go.dev/time#Timer.Reset

Is it necessary to reset the timer here? It's likely the timer hadn't fired yet.